internal/unionstore/pipelined_memdb.go

```diff
@@ -0,0 +1,343 @@
+	// MinFlushSize <= size < ForceFlushSizeThreshold && keys >= MinFlushKeys, flush.
+	// size >= ForceFlushSizeThreshold, flush.
+	if size < MinFlushSize || (p.memDB.Len() < MinFlushKeys && size < ForceFlushSizeThreshold) {
+		return false
+	}
+	if p.onFlushing.Load() && size < ForceFlushSizeThreshold {
+		return false
+	}
+	return true
+}
+
+// FlushWait will wait for all flushing tasks are done and return the error if there is a failure.
+func (p *PipelinedMemDB) FlushWait() error {
+	if p.flushingMemDB != nil {
+		err := <-p.errCh
+		// cleanup the flushingMemDB so the next call of FlushWait will not wait for the error channel.
+		p.flushingMemDB = nil
+		return err
+	}
+	return nil
+}
+
+// Iter implements the Retriever interface.
+func (p *PipelinedMemDB) Iter([]byte, []byte) (Iterator, error) {
+	return nil, errors.New("pipelined memdb does not support Iter")
+}
+
+// IterReverse implements the Retriever interface.
+func (p *PipelinedMemDB) IterReverse([]byte, []byte) (Iterator, error) {
+	return nil, errors.New("pipelined memdb does not support IterReverse")
+}
+
+// SetEntrySizeLimit sets the size limit for each entry and total buffer.
+func (p *PipelinedMemDB) SetEntrySizeLimit(entryLimit, bufferLimit uint64) {
+	p.entryLimit, p.bufferLimit = entryLimit, bufferLimit
+	p.memDB.SetEntrySizeLimit(entryLimit, bufferLimit)
+}
+
+func (p *PipelinedMemDB) Len() int {
+	return p.memDB.Len() + p.len
+}
+
+func (p *PipelinedMemDB) Size() int {
+	return p.memDB.Size() + p.size
+}
+
+func (p *PipelinedMemDB) OnFlushing() bool {
+	return p.onFlushing.Load()
+}
+
+// SetMemoryFootprintChangeHook sets the hook for memory footprint change.
+// TODO: implement this.
+func (p *PipelinedMemDB) SetMemoryFootprintChangeHook(hook func(uint64)) {}
+
+// Mem returns the memory usage of MemBuffer.
+// TODO: implement this.
+func (p *PipelinedMemDB) Mem() uint64 {
+	return 0
+}
+
+type errIterator struct {
+	err error
+}
+
+func (e *errIterator) Valid() bool   { return true }
+func (e *errIterator) Next() error   { return e.err }
+func (e *errIterator) Key() []byte   { return nil }
+func (e *errIterator) Value() []byte { return nil }
+func (e *errIterator) Close()        {}
+
+// SnapshotIter implements MemBuffer interface, returns an iterator which outputs error.
+func (p *PipelinedMemDB) SnapshotIter(k, upperBound []byte) Iterator {
+	return &errIterator{err: errors.New("SnapshotIter is not supported for PipelinedMemDB")}
+}
+
+// SnapshotIterReverse implements MemBuffer interface, returns an iterator which outputs error.
+func (p *PipelinedMemDB) SnapshotIterReverse(k, lowerBound []byte) Iterator {
+	return &errIterator{err: errors.New("SnapshotIter is not supported for PipelinedMemDB")}
+}
+
+// Reset the PipelinedMemDB to init state.
+func (p *PipelinedMemDB) Reset() {
+	p.memDB.Reset()
+	p.flushingMemDB = nil
+	p.len, p.size = 0, 0
+	p.onFlushing.Store(false)
+	p.generation = 0
+	// wait for the error channel to be empty.
```

Comment:
Should we wait for the flush to be finished first before the destruction? What are the use cases for `Reset`, when Reset is called and flush gets stuck, will it cause any issues?

Reply:
`Reset` should not be used IMO, because some flushed mutations can not be dropped by calling reset...

I don't see any usage of this function, what about just panic here?

Reply:
Or could we remove this function?

Reply:
I'll try, I find a test in TiDB will be affected need to be fixed.

Reply:
`Reset` is removed.